[優先度高]txBuilder.addSignatureで指定すべき引数を変更する必要がある

@KimuraYu45z 
ローカルでng serveしたところ、以下のようなエラーが発生しました。
cosmos-client-tsのバージョンをv0.42.4 -> v0.42.6に上げた際、txBuilder.addSignatureで指定すべき引数を変更する必要があるという状況かと見えています。
お手数ですが、上記コメント違和感ないかどうか確認いただけると助かります。
一旦、こちらで修正試みてみますが、解決に時間がかかりそうであれば、ご相談させてください。

```
Error: projects/cosmoscan/src/model/cosmos/bank.service.ts:62:37 - error TS2554: Expected 1 arguments, but got 2.

62     txBuilder.addSignature(privKey, signDoc);
                                       ~~~~~~~


Error: projects/cosmoscan/src/model/cosmos/staking.service.ts:67:37 - error TS2554: Expected 1 arguments, but got 2.

67     txBuilder.addSignature(privKey, signDoc);
```

@KimuraYu45z 
問題の箇所を以下のように置き換えたところ、

```
    const signDocBytes = txBuilder.signDocBytes(account.account_number);
    txBuilder.addSignature(privKey.sign(signDocBytes));
```

以下のようなエラーが発生しました。

```
Error: projects/cosmoscan/src/model/cosmos/bank.service.ts:58:49 - error TS2345: Argument of type 'Long | null | undefined' is not assignable to parameter of type 'Long'.
  Type 'undefined' is not assignable to type 'Long'.

58     const signDocBytes = txBuilder.signDocBytes(account.account_number);
```

cosmos-client-ts内で定義されているtxBuilder.signDocBytesの引数の型が、Longで定義されているのに対し、Long | null | undefinedな型の引数を指定しているのが問題だと思うのですが、昨夜加えて頂いた修正を考慮すると、ここで引数をLongに加工してはならず、nullならnullのまま、undefinedならnullにする(かあるいはundefinedのまま？)等して渡す必要があるという理解をしています。
したがって、cosmos-client-ts内で定義されているtxBuilder.signDocBytesに相当する関数の引数の型の定義を、Longではなく、Long | null | undefinedに修正し、nullやundefinedの場合も正常に動作するようにcosmos-client-ts側を修正する必要があるという理解をしているのですが正しいでしょうか？

まだ、cosmos-client-ts側をチェックしていないのですが、上記理解が正しく、こちらでも修正可能なレベル感であれば、cosmos-client-ts側の修正をトライしてみようと思います。

上記ご確認の上、こちらでcosmos-client-ts側の修正を行ったほうが良いか？、あるいは木村さんにて修正頂いたほうが良いか？あるいは何か見当違いのことを言っていないか等、アドバイス頂けると助かります。

よろしくお願いします。

--- a/package.json
+++ b/package.json
@@ -27,7 +27,7 @@
     "@angular/service-worker": "~12.1.1",
     "@types/bip39": "^3.0.0",
     "bip39": "^3.0.3",
-    "cosmos-client": "^0.42.6",
+    "cosmos-client": "^0.42.7",
     "dexie": "^3.0.2",
     "dotenv": "^8.2.0",
     "ng-loading-dialog": "^12.0.5",

--- a/projects/cosmoscan/src/model/cosmos/bank.service.ts
+++ b/projects/cosmoscan/src/model/cosmos/bank.service.ts
@@ -8,13 +8,10 @@ import { KeyService } from '../keys/key.service';
   providedIn: 'root',
 })
 export class BankService {
-  constructor(
-    private readonly cosmosSDK: CosmosSDKService,
-    private readonly key: KeyService,
-  ) { }
+  constructor(private readonly cosmosSDK: CosmosSDKService, private readonly key: KeyService) { }
 
   async send(key: Key, toAddress: string, amount: cosmos.base.v1beta1.ICoin[], privateKey: string) {
-    const sdk = await this.cosmosSDK.sdk().then(sdk => sdk.rest)
+    const sdk = await this.cosmosSDK.sdk().then((sdk) => sdk.rest);
     const privKey = this.key.getPrivKey(key.type, privateKey);
     const pubKey = privKey.pubKey();
     const fromAddress = cosmosclient.AccAddress.fromPublicKey(pubKey);
@@ -22,18 +19,18 @@ export class BankService {
     // get account info
     const account = await rest.cosmos.auth
       .account(sdk, fromAddress)
-      .then((res) => res.data.account && cosmosclient.codec.unpackAny(res.data.account) as cosmos.auth.v1beta1.IBaseAccount)
+      .then((res) => res.data.account && (cosmosclient.codec.unpackAny(res.data.account) as cosmos.auth.v1beta1.IBaseAccount))
       .catch((_) => undefined);
 
     if (!account) {
-      throw Error('Address not found')
+      throw Error('Address not found');
     }
 
     // build tx
     const msgSend = new cosmos.bank.v1beta1.MsgSend({
       from_address: fromAddress.toString(),
       to_address: toAddress,
-      amount: amount,
+      amount,
     });
 
     const txBody = new cosmos.tx.v1beta1.TxBody({

--- a/projects/cosmoscan/src/model/cosmos/bank.service.ts
+++ b/projects/cosmoscan/src/model/cosmos/bank.service.ts
@@ -19,10 +19,10 @@ export class BankService {
     // get account info
     const account = await rest.cosmos.auth
       .account(sdk, fromAddress)
-      .then((res) => res.data.account && (cosmosclient.codec.unpackAny(res.data.account) as cosmos.auth.v1beta1.IBaseAccount))
+      .then((res) => res.data.account && cosmosclient.codec.unpackAny(res.data.account))
       .catch((_) => undefined);
 
-    if (!account) {
+    if (!(account instanceof cosmos.auth.v1beta1.BaseAccount)) {
       throw Error('Address not found');
     }
 
@@ -55,8 +55,8 @@ export class BankService {
 
     // sign
     const txBuilder = new cosmosclient.TxBuilder(sdk, txBody, authInfo);
-    const signDoc = txBuilder.signDoc(account.account_number!);
-    txBuilder.addSignature(privKey, signDoc);
+    const signDocBytes = txBuilder.signDocBytes(account.account_number);
+    txBuilder.addSignature(privKey.sign(signDocBytes));
 
     // broadcast
     const result = await rest.cosmos.tx.broadcastTx(sdk, {

--- a/projects/cosmoscan/src/model/cosmos/staking.service.ts
+++ b/projects/cosmoscan/src/model/cosmos/staking.service.ts
@@ -19,10 +19,10 @@ export class StakingService {
     // get account info
     const account = await rest.cosmos.auth
       .account(sdk, fromAddress)
-      .then((res) => res.data.account && (cosmosclient.codec.unpackAny(res.data.account) as cosmos.auth.v1beta1.IBaseAccount))
+      .then((res) => res.data.account && (cosmosclient.codec.unpackAny(res.data.account) as cosmos.auth.v1beta1.BaseAccount))
       .catch((_) => undefined);
 
-    if (!account) {
+    if (!(account instanceof cosmos.auth.v1beta1.BaseAccount)) {
       throw Error('Address not found');
     }
 
@@ -55,8 +55,8 @@ export class StakingService {
 
     // sign
     const txBuilder = new cosmosclient.TxBuilder(sdk, txBody, authInfo);
-    const signDoc = txBuilder.signDoc(account.account_number!);
-    txBuilder.addSignature(privKey, signDoc);
+    const signDocBytes = txBuilder.signDocBytes(account.account_number);
+    txBuilder.addSignature(privKey.sign(signDocBytes));
 
     // broadcast
     const result = await rest.cosmos.tx.broadcastTx(sdk, {

--- a/projects/cosmoscan/src/model/cosmos/staking.service.ts
+++ b/projects/cosmoscan/src/model/cosmos/staking.service.ts
@@ -8,18 +8,10 @@ import { cosmosclient, rest, cosmos } from 'cosmos-client';
   providedIn: 'root',
 })
 export class StakingService {
-  constructor(
-    private readonly cosmosSDK: CosmosSDKService,
-    private readonly key: KeyService,
-  ) { }
+  constructor(private readonly cosmosSDK: CosmosSDKService, private readonly key: KeyService) { }
 
-  async createDelegator(
-    key: Key,
-    validatorAddress: string,
-    amount: cosmos.base.v1beta1.ICoin,
-    privateKey: string,
-  ) {
-    const sdk = await this.cosmosSDK.sdk().then(sdk => sdk.rest)
+  async createDelegator(key: Key, validatorAddress: string, amount: cosmos.base.v1beta1.ICoin, privateKey: string) {
+    const sdk = await this.cosmosSDK.sdk().then((sdk) => sdk.rest);
     const privKey = this.key.getPrivKey(key.type, privateKey);
     const pubKey = privKey.pubKey();
     const fromAddress = cosmosclient.AccAddress.fromPublicKey(pubKey);
@@ -27,18 +19,18 @@ export class StakingService {
     // get account info
     const account = await rest.cosmos.auth
       .account(sdk, fromAddress)
-      .then((res) => res.data.account && cosmosclient.codec.unpackAny(res.data.account) as cosmos.auth.v1beta1.IBaseAccount)
+      .then((res) => res.data.account && (cosmosclient.codec.unpackAny(res.data.account) as cosmos.auth.v1beta1.IBaseAccount))
       .catch((_) => undefined);
 
     if (!account) {
-      throw Error('Address not found')
+      throw Error('Address not found');
     }
 
     // build tx
     const msgDelegate = new cosmos.staking.v1beta1.MsgDelegate({
       delegator_address: fromAddress.toString(),
       validator_address: validatorAddress,
-      amount: amount,
+      amount,
     });
 
     const txBody = new cosmos.tx.v1beta1.TxBody({

